### ISSUE TITLE
Reconnect doesn't seem to work with dotnet watch

### ISSUE THREAD

**Issue by rynowak** [aspnet/AspNetCore#9256]
If I launch a components app with `dotnet watch run` - the reconnection feature doesn't ever seem to do its thing.

![image](https://user-images.githubusercontent.com/1430011/55903879-713a4b80-5b83-11e9-8e32-7acdd1ccaa3f.png)

I can retry, and it will successfully connect (see console output) - but it never updates the UI.

My guess is that this happens because the circuit is gone - it's a new process after all. We should see if we can improve this for inner loopness, and figure out a better experience by default for the case where the circuit is gone.

**Comment by rynowak** [comment]
@javiercn in case he has cool thoughts

**Comment by javiercn** [comment]
Yeah, my guess is the same as yours. The only option at that point is to F5 the app. We should inform the client that the circuit is gone and provide some button on the UI to refresh the page as at that point the app is gone.

**Comment by javiercn** [comment]
I think this not only applies to developer inner loop. In the future we might want to free up resources in case the app is idle for a while, at that point there should be some integration so that developers can take that into account and provide a good experience.

For example, something like
```
The application has been inactive for too long so we have ended your current session. 
Click here to refresh the application.
```

**Comment by danroth27** [comment]
We discussed this and we think it would help that if the circuit ID is unknown we reload the page.

**Comment by danroth27** [comment]
We think we should make auto refreshing the browser by default and provide an option to prompt the user to do it instead, like we do today. We could enable this option during development so that the error state is not lost.

**Comment by gingters** [comment]
I think that behavior should be configurable. In development mode a reload is fine and most likely expected.

In production, there could be a lot of issues (i.e. server node accidentally or gracefully restarted) and that might involve trying to persist some sort of session state when the server gracefully shuts down and trying to restore that state when a new circuit starts.

Also, the user might want to be informed when he should expect an application reload and possibly loss of data.

**Comment by barahonajmc** [comment]
What I found:

### dotnet watch reconnection
This works:
1. Create a new blazor server app
2. Run it with `dotnet watch run`
3. Change some files
4. The restart and reconnection of the server works fine

### Restart the server manually
1.- Manually restart the server by killing the process and then `dotnet watch run` again
2.- Old pages will try to reconnect until they notice the server is no longer the same and will display the following:
![image](https://user-images.githubusercontent.com/15148088/95469326-872df580-094d-11eb-994e-acc3da062ee1.png)


So, unless I am missing something, this is already working, right?

**Comment by captainsafia** [comment]
@barahonajmc Thanks for testing it out! Can you confirm what version you validated this on?

**Comment by barahonajmc** [comment]
oh I forget that, it was on .NET 5 RC1

**Comment by BieleckiLtd** [comment]
Here is how the user experience looks in my case.

User loads page, works with the app, perhaps fills the form halfway through. They go for a snack and when they come back they see that their work is gone as the only thing they can do is to refresh the browser. Great for the banking app, but my users complain.

Disabling idle timeout would help but app pool can still recycle at any time unless this is also disabled. Corporate IT is not happy to change any of the application pool settings anyway. Similarly with my hobby stuff - online hosting providers are saying to go and buy VPS if I need such customization.

tl;dr
I know this is a long shot but are there any plans to improve user experience with blazor after the application restart? Maybe persist user session state in the browser?

**Comment by konradbartecki** [comment]
I actually implemented the exact scenario where this is configurable, but with default `reload page on circuit rejected`
See: #32122

**Comment by TanayParikh** [comment]
Duplicated https://github.com/dotnet/aspnetcore/issues/38305 against this issue. Thread summary is there are concerns with the [existing documented approach](https://docs.microsoft.com/en-us/aspnet/core/blazor/fundamentals/signalr?view=aspnetcore-6.0#modify-the-reconnection-handler-blazor-server) as it overrides the existing reconnection behavior. 

We may want to consider adding additional doc details on the reconnection behavior or provide alternate reconnection configurability (or automatic reloads).

**Comment by TanayParikh** [comment]
Related:

- https://github.com/dotnet/aspnetcore/issues/44669
- https://github.com/dotnet/aspnetcore/issues/43513
- https://github.com/dotnet/aspnetcore/issues/39409
- https://github.com/dotnet/aspnetcore/issues/30344
- https://github.com/dotnet/aspnetcore/issues/30287

**Comment by danroth27** [comment]
We're planning to address this as part of https://github.com/dotnet/aspnetcore/issues/9256

**Comment by aryehsilver** [comment]
> We're planning to address this as part of #9256

Which is this issue but has now been closed?!?

**Comment by danroth27** [comment]
Dah! Sorry! The correct tracking issue is #32113.
